Handle API JSON format

schmidt-sebastian opened the pull request:
This adds API JSON format to our internal snapshot_() API as specified by go/api-json (Sorry, this is Google Internal). This format is used by Firebase Functions, and directly exposing it means that the GCF team does not have to worry about our data format.

This was added as a conversion (rather than just adding it to decodeValue), since the logic in watch.js and order.js relies on the Protobuf JS format.

We can potentially get rid of creating the extra copy of the data, but that would mean that the input to snapshot_() gets modified.

mikelehen commented on test/index.js:
```diff
@@ -291,6 +297,96 @@ describe('instantiation', function() {
+    assert.equal('1970-01-01T00:00:05.000000006Z', doc.readTime);
+  });
+
+  it('handles API JSON', function() {
```
Maybe add a comment something like:

// Google Cloud Functions must be able to call snapshot_() with API JSON data.

mikelehen commented on src/index.js:
```diff
@@ -351,21 +355,24 @@ class Firestore extends commonGrpc.Service {
       document.ref = new DocumentReference(
         this,
         ResourcePath.fromSlashSeparatedString(documentOrName.name)
       );
```
Maybe add a comment something like:

// Cloud Functions calls us with API JSON format data, so we must be sure to convert it.

schmidt-sebastian commented on src/convert.js:
```diff
@@ -0,0 +1,222 @@
+
+    if (isNaN(seconds) || isNaN(nanos)) {
+      // This error should only ever be thrown if the end-user specifies an
+      // invalid 'lastUpdateTime'.
```
This comment is meant to explain why I use "lastUpdateTime" and not "timestampValue" in the error message. I took this opportunity to expand on the comment since it obviously didn't quite convey this meaning :)

mikelehen replied:
Oh, I see!  This seems a bit fragile...  I'd expect to either pass in the parameter name as a string to use in the error message or to throw a generic error and have the calling code try/catch it and throw a more friendly error.  I'm fine with this as-is, but I'd be tempted to add "MINOR HACK: " or similar to the comment to make it clearer that we're taking a shortcut here...

schmidt-sebastian commented on src/convert.js:
```diff
@@ -0,0 +1,222 @@
+  }
+}
+/**
+ * Converts an API JSON or Protobuf JS 'Document' into Protobuf JS. This
```
Done (here and everywhere else)

schmidt-sebastian commented on src/document.js:
```diff
@@ -1053,14 +1054,11 @@ class Precondition {
-        seconds: seconds,
-        nanos: nanos,
-      };
+      proto.updateTime = convertTimestamp(this._lastUpdateTime);
```
Ugh, that was kind of the point of this PR. Sorry that I missed that.

schmidt-sebastian commented on src/convert.js:
```diff
@@ -0,0 +1,222 @@
+}
+/**
+ * Converts an API JSON or Protobuf JS 'Document' into Protobuf JS. This
+ * conversion creates a copy of underlying document.
```
Done

schmidt-sebastian commented:
> I think this basically looks good, though I'd like to better capture what's going on here in code comments.
> 
> At minimum, I think convert.js could maybe use a file-level comment explaining "Protobuf JS" and "API JSON" and mentioning the relation to Cloud Functions.
> 
> Also, is there any better way to describe "API JSON" ? Searching for that internally or externally doesn't yield anything useful, so I'm not sure how folks will understand this code. :-( Should it be "proto3 JSON" or something?

I did think about adding a module level comment before, but I didn't see it anywhere else in the codebase, so I omitted it at first. I made it private and used "/*!" so I think it won't show up in the generated docs.

I also changed the way the encoding works to only only convert the JSON format when required (snapshot_ now takes an encoding format). I replaced all mentions of "API JSON" to "Proto3 JSON". This does seem more common (used by https://developers.google.com/protocol-buffers/docs/proto3#json), and I hope that this is the right term - the document that describes the format doesn't mention a public name.

schmidt-sebastian commented on src/index.js:
```diff
@@ -336,19 +333,41 @@ class Firestore extends commonGrpc.Service {
   /**
    * Creates a [DocumentSnapshot]{@link DocumentSnapshot} from a
-   * `Document` proto (or from a resource name for missing documents).
+   * `firestore.v1beta1lDocument` proto (or from a resource name for missing
```
Fixed.

schmidt-sebastian commented:
> Oh, I see! This seems a bit fragile... I'd expect to either pass in the parameter name as a string to use in the error message or to throw a generic error and have the calling code try/catch it and throw a more friendly error. I'm fine with this as-is, but I'd be tempted to add "MINOR HACK: " or similar to the comment to make it clearer that we're taking a shortcut here...

Updated to pass in the argument name (it's only used in a few places).

schmidt-sebastian commented:
@lukesneeringer @stephenplusplus Added you for final Approval. Thanks.
